### ISSUE TITLE
[OEP] OSeMOSYS Hydro

### ISSUE THREAD

**Issue by MonacoL** [OSeMOSYS/OSeMOSYS#98]
### What are you proposing to add to the OSeMOSYS formulation?
"OSeMOSYS Hydro" contributes a novel integrated energy-water system model where hydropower reservoirs planning and management is optimized jointly with energy system capacity expansion and generation mix. It can be seen as a modular extension of the well known open source energy system optimization model "OSeMOSYS" based on "OSeMOSYS for Cascaded Hydro Code" developed by T. Niet. It is made with the Pyomo version of the model.
In "OSeMOSYS Hydro", reservoirs are hydraulically linked as the rivers and their tributaries flow through them using different lag-times between reservoirs. In addition to that, depending on the actual volume of the reservoir, different constraints for maximum and minimum release, net evaporation flows are enforced. Last, rule curves are used to define a regulation zone where reservoir management decision has to be optimized. 
To reduce the computational cost of such new features, this extension provides the opportunity to use a fix-and-relax approach during the optimization. "OSeMOSYS Hydro" defines a MILP, while pure OSeMOSYS defines a LP. During the fix-and-relax optimization, both are used. Given a certain time horizon, the final water-energy plan is found recursively. In the first simulation, OSeMOSYS Hydro is performed on a limited window while OSeMOSYS is performed in the remaining years. This happens during the same optimization. In the next simulations the window is shifted towards the future, the old window's results are fixed and OSeMOSYS is applied on the remaining years (that grow smaller every subsequent simulation). On the last simulation, the window reaches the end of the time horizon, all the previous years are fixed and there are no more following years to apply OSeMOSYS.
This approach approximates the best solution, which would be given by performing OSeMOSYS Hydro on the whole time horizon with one simulation, but it is necessary in order to obtain a reasonable water-energy plan in few hours with normally capable servers.
One can decide to perform a pure OSeMOSYS Hydro optimization with no fix-and-relax, by setting the length of the window as the length of the time horizon.

### Why is this required?
Up-to-date modelling tools holistically describing the water-energy nexus are required to provide a comprehensive prospective to support planning processes. Indeed, existing energy system models generally include overly simplified characterisation of the water component. This work contributes a new improved energy-water modelling scheme, which includes the main components of the hydrological cycle and their dynamics into the energy system optimization.

### Detailed description of change
In OSeMOSYS Hydro, power plants are described as technologies. Reservoirs are implemented as completely new sets.
1. addition, removal or changes of sets, parameters
NEW SETS:
ActiveYears: set of total years to be optimized (MILP+LP YEARS)
ActiveMILPYears: set of MILP years to be optimized
FixedMILPYears: set of MILP years fixed from previous simulations
Reservoir: set of all the reservoirs taken into account, both constructed (MaxNewCapacity zero) and yet to be constructed (MaxNewCapacity non zero)
I,J: RangeSet for SOS constraints

NEW PARAMETERS:
RhoWater: water density in kg/m3
Gravity: gravity acceleration in m/s2
FlowUnitToYearConversion, WattsToModelUnits: conversion parameters useful for both constraints and objective function
HydroTurbineEfficiency: it ranges from 0 to 1, it defines how much efficient the turbines are for each power plants
TechnologyFromHydro: it is binary, it relates a reservoir to its powerplants
ReservoirHead: per each reservoir. Given that the problem is linear, it is operationally an average head, in km3
ReservoirStorageVolume: volume of the reservoir, km3
MinimumEnvironmentalFlow: minimum release of a reservoir that has to occur no matter the conditions, m3/s
ReservoirLevelStart: reservoir's volume at the first month of the first year of the optimization, in m3
StartRelease: it defines in which month of the year a reservoir can begin to release water. 
ReservoirExternalInflow: external inflow from the basins to each reservoir, in m3/s
Evap: evaporation, per each month, per each reservoir, in mm/year
DownstreamReservoirTag: it links subsequent reservoirs, from an upstream one to a downstream one. It manages to link multiple upstream reservoir to the same downstream reservoir.
InflowLagtime: it defines the delay in month, between the release of upstream water upstream and its arrival to a downstream reservoir
RuleCurve: ideal storage trajectories over time, per each reservoir, that the optimization takes as a benchmark, in km3
ROR: RunOfRiver, it specifies if a reservoir has a run of river hydro power plant
BIGM, BIGM2: two big values used for a BIGM formulation of some constraints

Parameters that take the value of the output variables of the previous simulation are defined. They are used in the objective function and they are defined for the set FIXEDMILPYEARS.

NEW VARIABLES:
RateOfReservoirDischarge: part of a reservoir's outward flow that is used to produce energy, in m3/s
RateOfReservoirSpillage: part of a reservoir's outward flow that is NOT used to produce energy, in m3/s
Release: a reservoir's total outward flow, in m3/s
ReservoirLevelSeason: reservoir's storage, per each month, in km3
Inflow: sum of external inflow to a reservoir and downstream release from an uphill reservoir
ReservoirLevelFinal: reservoir's storage at the end of the optimization
x,y: positive (x) and negative (y) difference between a reservoir's storage and its rule curve value, per each month, in km3 
Surface: water surface related to a reservoir's storage, in km2

The model uses maximum and minimum release matrices to define max and min release values per each reservoir, in function of monthly inflow and monthly storage. In order to fetch the data, SOS constraints are used.

beta, alpha: binary variables used in the SOS constraints to select the right cell from the matrices
MaxRel, MinRel: the selected values of maximum release and minimum release, per each month, per each reservoir
CostruiteBIGM: binary variable, per each yet to be constructed reservoir, per each MILPYEAR, that specify if and when a hydropower plant is built

2. addition, removal or changes of equations
The objective function of OSeMOSYS is used with a hydrological addition. In OSeMOSYS Hydro a reservoir's storage difference from its rule curve (variables x and y) is translated in an overproduction or an underproduction of energy (depends if the difference is positive or negative) by the associated hydropower plants, basically giving the (production) cost related to the storage difference from the rule curve. This cost is added to the objective function and so it is intended to be minimized. In this way, the storages are optimized.

3. addition, remove or changes of constraints
Added a constraint to compute hydroproduction using hydropower production law, which is linearized by using an average reservoir head per each reservoir. This defines the variable RateOfReservoirDischarge.
Added a constraint to compute the storage balance between months, per each reservoir, setting the storage at the current month using the previous month storage, the total inflow at the current month, the release at the current month and the evaporation from the surface of the storage at the current month.
Added some SOS constraints to select the max and min release values and some to impose those values.
Added some constraints to make sure that the storage at a certain month would not exceed the maximum volume of the reservoir.
Added 2 contraints to manage the construction of the new hydropower plants as linked to their reservoirs, with a BIGM formulation.
Added many constraints that manages the release of the new reservoirs ONLY only after the construction of their hydropower plants. Before, new reservoirs are non-operational, giving outward release = inward inflow per each month.
Added many constraint to optimize the storages with rule curves ONLY only after the construction of their hydropower plants.
For every reservoir, a relation volume-surface is defined for a few points that ranges from empty reservoir to full reservoir condition. It is important that the surface is computed for every monthly storage following the given the relation, in order to compute correctly the monthly evaporation (which is intended to be per unit of surface) to be used in the storage balance constraint. This has been achieved by using a Piecewise constraint.

**Comment by willu47** [comment]
Dear @MonacoL - many thanks for the excellent and comprehensive submission to the (brand new) OEP process.  Please note that we (@abhishek0208, @tniet and I) as representatives of the wider OSeMOSYS community) would like to use your submission as a test-case (if that's okay) to better develop this process.

The key steps of this (in-development) process are:

1. initial review - which seeks answers to questions like: does this fit within the aims of OSeMOSYS? Will the community find this useful? Is the OEP well developed? How long will it take to implement? What are the implications for the different implementations of OSeMOSYS?
2. Implementation - in which the work of doing the coding/restructing/documentation etc. takes place
3. Acceptance review - final review of the various components of the change

Note that the result of 1 or 3 may be that the OEP is rejected, or that a resubmission is required.

So, next steps, I'll invite @abhishek0208, @tniet and others to offer their comments below for the *initial review*.

**Comment by MonacoL** [comment]
Dear Prof. Usher,
of course you can use this submission as a test-case.

Waiting for your next comments.

Luca

Il giorno mar 22 ott 2019 alle ore 22:14 Will Usher <
notifications@github.com> ha scritto:

**Comment by JZF07** [comment]
Dear Luca and Will, 

This is an interesting osemosys-application and it has been tackled in previous years in different ways. one humble example is given in this study https://journals.aau.dk/index.php/sepm/article/view/2156/2023 
where a simplified hydrological system was introduced to osemosys to capture the some operation dynamics (i.e. cooperation versus no cooperation) between HPPs in the cascade. 
Best, 
Youssef

**Comment by tniet** [comment]
Hi Luca,

Thanks for the submission, and for taking the Hydro code I published and expanding/improving on it!

My comments at this point, for initial review.  I think it's a well developed OEP and the implementation seems to be well on the way.  The main concern/question I have is should this be part of the core OSeMOSYS code or should it be kept as a 'branch' that others can use if they feel it fits within the type of analysis they want to do.  With the original Hydro code you reference it was purposefully not included in the main OSeMOSYS code, so I think this falls into the same category.

I'm also concerned about the MILP implications and how this impacts runtime.  Again, maybe it's not part of the official release version but ends up in a list of interesting code additions that can be used if needed.

Cheers,
Taco..

**Comment by MonacoL** [comment]
Dear Prof. Niet
I agree on seeing this oep as an extension/branch of the main OSeMOSYS, so
that anyone could choose to use it only if needed. Actually, It was meant
to be so.
Regarding MILP implications, I am here for every detailed questions (but I
think we Will have the chance to discuss the matter when we land to the
second phase of the oep... IF we land there)

Luca

Luca

**Comment by MonacoL** [comment]
Dear All,
I haven't heard from you in a while.
Is this OEP meant to proceed?

**Comment by willu47** [comment]
Hi @MonacoL - given that this will live as a branch on the Pyomo version of OSeMOSYS, I think you can go ahead an make a pull request there including your changes.

Please be sure to follow the correct workflow for uploading your changes.

1. [Fork](https://github.com/OSeMOSYS/OSeMOSYS_Pyomo/fork) the [OSeMOSYS_Pyomo](https://github.com/osemosys/OseMOSYS_Pyomo) repository to your own Github user account
2. Create a new branch
3. Add and commit your modifications to the original OSeMOSYS_Pyomo code (do not add new files - you should replace the content of the original files with your modifications)
4. Add documentation files and update the readme
5. Push your new branch to your forked copy of the repository
6. Submit a pull request from your forked copy of the repository to a new branch on the  OSeMOSYS/OSeMOSYS_Pyomo repository referencing this OEP using #98

Further comments and suggestions can happen on the pull request.
